File: akka-mapreduce-java/src/main/java/cn/v5cn/akka/mapreduce_java/actors/AggregateActor.java
package cn.v5cn.akka.mapreduce_java.actors;

import java.util.HashMap;
import java.util.Map;

import akka.actor.ActorRef;
import akka.actor.UntypedActor;
import cn.v5cn.akka.mapreduce_java.messages.ReduceData;
import cn.v5cn.akka.mapreduce_java.messages.Result;

public class AggregateActor extends UntypedActor {

	private Map<String,Integer> finalReducedMap = new HashMap<String,Integer>();
	
	@Override
	public void onReceive(Object message) throws Exception {
		if(message instanceof ReduceData){
			ReduceData reduceData = (ReduceData)message;
			aggregateInMemoryReduce(reduceData.getReduceDataList());
		}else if(message instanceof Result){
			getSender().tell(finalReducedMap.toString(),ActorRef.noSender());
		}else{
			unhandled(message);
		}
	}

	private void aggregateInMemoryReduce(HashMap<String, Integer> reduceDataList) {
		Integer count = null;
		for(String key : reduceDataList.keySet()){
			if(finalReducedMap.containsKey(key)){
				count = reduceDataList.get(key) + finalReducedMap.get(key);
				finalReducedMap.put(key, count);
			}else{
				finalReducedMap.put(key, reduceDataList.get(key));
			}
		}
	}

}

File: akka-mapreduce-scala/src/main/scala/cn/v5cn/akka/mapreduce_scala/actors/AggregateActor.scala
package cn.v5cn.akka.mapreduce_scala.actors

import akka.actor.Actor
import cn.v5cn.akka.mapreduce_scala.ReduceData
import cn.v5cn.akka.mapreduce_scala.Result
import scala.collection.mutable.HashMap

class AggregateActor extends Actor {
  
  val finalReducedMap = new HashMap[String ,Int]
  
	def receive = {
	  case ReduceData(reduceDataMap) =>
	    aggregateInMemoryReduce(reduceDataMap)
	  case Result =>
	    sender ! finalReducedMap.toString()
	}
  
  def aggregateInMemoryReduce(reducedList:Map[String,Int]):Unit={
    for((key,value) <- reducedList){
      if(finalReducedMap contains key){
        finalReducedMap(key) = (finalReducedMap.get(key).get + value)
      }else{
        finalReducedMap += (key -> value)
      }
    }
  }
}